clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4643,8 +4654,11 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     if (!IsSYCLOffloadDevice) {
       // Add the -include option to add the integration header
       StringRef Header = D.getIntegrationHeader(Input.getBaseInput());
+      // Do not add the integration header if we are compiling after the
+      // integration footer has been applied.  Check for the append job
+      // action to determine this.
       if (types::getPreprocessedType(Input.getType()) != types::TY_INVALID &&
-          !Header.empty()) {
+          !Header.empty() && !ContainsAppendFooterAction(&JA)) {
```

Comment:
Would it be possible to check if we're working on a preprocessing `JobAction` instead? I have vague suspicions that this might fail for the footer-absent cases with Clang-based actions collapsed, but a confirmation/denial of this would be of great help.

Reply:
I don't think checking for a preprocessing action is enough here.  For host compilation and preprocessing, we want to add the integration header (regular host compilation flow)  We need to be able to differentiate the initial host preprocessing step and the later preprocessing step that occurs with the integration footer and only pull in the integration header for the flow which does not include the footer.  Can you elaborate on your suspicions of compilations that will not have the footer?  When the toolchain is put together, the footer call is always done, regardless if there are any contents in the footer or not.

Reply:
> Can you elaborate on your suspicions of compilations that will not have the footer?

I was implying potential issues that could arise if the check was switched to the one for preprocess action. However:
> When the toolchain is put together, the footer call is always done, regardless if there are any contents in the footer or not.

That's what I missed out on.

> We need to be able to differentiate the initial host preprocessing step and the later preprocessing step that occurs with the integration footer

...as well as this. My initial comment can therefore be disregarded; thanks for the clarification!